File: backend/state.py
import threading
from datetime import datetime

DISTRACTOR_WEIGHTS = {
    "microsleep":      10,
    "phone_check":      5,
    "disallowed_tab":   4,
    "yawn":             3,
    "tab_switch":       2,
    "eyes_off_screen":  2,
    "rewind":           2,
    "head_tilt":        1,
}

class SessionState:
    def __init__(self):
        self._lock = threading.RLock()
        self.reset()

    def reset(self):
        with self._lock:
            self.session_id = None
            self.is_active = False
            self.started_at = None
            self.counts: dict[str, int] = {k: 0 for k in DISTRACTOR_WEIGHTS}
            self.focus_score: float = 100.0
            # Latest CV metrics (updated every frame)
            self.ear: float = 0.3
            self.mar: float = 0.1
            self.head_tilt: float = 0.0
            self.nose_ratio: float = 0.3
            self.face_detected: bool = False
            self.blink_count: int = 0
            self.blink_rate: float = 0.0
            # Focus time accumulator (seconds face was detected during session)
            self.focus_seconds: float = 0.0
            # Calibration
            self.is_calibrating: bool = False
            self.calibrated: bool = False
            self._cal_nose: list[float] = []
            self._cal_tilt: list[float] = []
            self.nose_baseline: float = 0.5
            self.tilt_baseline: float = 0.0

    def start(self, session_id: str):
        with self._lock:
            self.session_id = session_id
            self.is_active = True
            self.started_at = datetime.utcnow()
            self.counts = {k: 0 for k in DISTRACTOR_WEIGHTS}
            self.focus_score = 100.0

    def stop(self):
        with self._lock:
            self.is_active = False

    def increment(self, event_type: str):
        with self._lock:
            if event_type in self.counts:
                self.counts[event_type] += 1
            self._recalculate()

    def _recalculate(self):
        deducted = sum(
            self.counts[k] * DISTRACTOR_WEIGHTS[k]
            for k in self.counts
        )
        self.focus_score = max(0.0, 100.0 - deducted)

    def top_distractors(self, n: int = 5):
        with self._lock:
            ranked = sorted(
                ((k, v) for k, v in self.counts.items() if v > 0),
                key=lambda x: x[1] * DISTRACTOR_WEIGHTS[x[0]],
                reverse=True,
            )
            return ranked[:n]

    def start_calibration(self):
        with self._lock:
            self.is_calibrating = True
            self.calibrated = False
            self._cal_nose = []
            self._cal_tilt = []

    def add_calibration_frame(self, nose: float, tilt: float):
        with self._lock:
            if self.is_calibrating:
                self._cal_nose.append(nose)
                self._cal_tilt.append(tilt)

    def finish_calibration(self):
        import statistics
        with self._lock:
            if self._cal_nose:
                self.nose_baseline = statistics.median(self._cal_nose)
                self.tilt_baseline = statistics.median(self._cal_tilt)
            self.is_calibrating = False
            self.calibrated = True

    def add_focus_time(self, dt: float):
        with self._lock:
            self.focus_seconds += dt

    def push_browser_focus(self, face_detected: bool, focus_score: float):
        """Called by the browser's MediaPipe pipeline so the desktop overlay stays in sync."""
        with self._lock:
            self.face_detected = face_detected
            self.focus_score = max(0.0, min(100.0, focus_score))

    def update_cv(self, *, ear, mar, head_tilt, nose_ratio, face_detected, blink_rate=0.0):
        with self._lock:
            self.ear = ear
            self.mar = mar
            self.head_tilt = head_tilt
            self.nose_ratio = nose_ratio
            self.face_detected = face_detected
            self.blink_rate = blink_rate

    def snapshot(self) -> dict:
        with self._lock:
            return {
                "session_id": self.session_id,
                "is_active": self.is_active,
                "focus_score": round(self.focus_score, 2),
                "counts": dict(self.counts),
                "top_distractors": self.top_distractors(5),
                "ear": round(self.ear, 4),
                "mar": round(self.mar, 4),
                "head_tilt": round(self.head_tilt, 2),
                "nose_ratio": round(self.nose_ratio, 4),
                "face_detected": self.face_detected,
                "blink_rate": round(self.blink_rate, 2),
                "focus_seconds": round(self.focus_seconds, 1),
                "calibrated": self.calibrated,
                "is_calibrating": self.is_calibrating,
                "nose_baseline": round(self.nose_baseline, 4),
                "tilt_baseline": round(self.tilt_baseline, 2),
            }


state = SessionState()

File: backend/lib/state.js
// In-memory session state — single instance shared across all routes.
// Updated by:
//   • Browser MediaPipe  →  POST /state/push  { face_detected, focus_score }
//   • Python CV bridge   →  POST /cv/push     { ear, mar, head_tilt, … }
//   • Event logging      →  POST /events      (recordEvent increments counts)

export const WEIGHTS = {
  microsleep:      10,
  phone_check:      5,
  yawn:             3,
  tab_switch:       2,
  eyes_off_screen:  2,
  head_tilt:        1,
};

export const TIPS = {
  microsleep:      "Take a 5-minute break every 25 minutes and ensure you're getting enough sleep.",
  phone_check:     "Put your phone face-down or in another room during focus sessions.",
  yawn:            "Stay hydrated and consider a quick walk to boost alertness.",
  tab_switch:      "Group all your research before starting — minimize context switching.",
  eyes_off_screen: "Position your monitor at eye level to reduce neck strain.",
  head_tilt:       "Adjust your monitor height so you look slightly downward.",
};

class SessionState {
  constructor() { this.reset(); }

  reset() {
    this.sessionId    = null;
    this.isActive     = false;
    this.startedAt    = null;
    this.counts       = Object.fromEntries(Object.keys(WEIGHTS).map(k => [k, 0]));
    this.focusScore   = 100;
    this.focusSeconds = 0;
    // Raw CV metrics (populated by Python bridge)
    this.ear          = 0.3;
    this.mar          = 0.1;
    this.headTilt     = 0;
    this.noseRatio    = 0.3;
    this.faceDetected = false;
    this.blinkRate    = 0;
    // Session config
    this.sessionType  = 'general';
    this.allowedTabs  = [];
    // Internal timing for focus-seconds accumulation
    this._lastPushMs  = null;
  }

  start(sessionId, { session_type = 'general', allowed_tabs = [] } = {}) {
    this.reset();
    this.sessionId   = sessionId;
    this.isActive    = true;
    this.startedAt   = new Date();
    this.sessionType = session_type;
    this.allowedTabs = allowed_tabs;
    this._lastPushMs = Date.now();
  }

  stop() { this.isActive = false; }

  // Called by POST /events — increments count and lowers score
  recordEvent(type) {
    if (type in this.counts) this.counts[type]++;
    this._recalculate();
  }

  _recalculate() {
    const deducted = Object.entries(this.counts)
      .reduce((sum, [k, v]) => sum + v * (WEIGHTS[k] ?? 0), 0);
    this.focusScore = Math.max(0, 100 - deducted);
  }

  topDistractors(n = 5) {
    return Object.entries(this.counts)
      .filter(([, v]) => v > 0)
      .sort((a, b) =>
        b[1] * (WEIGHTS[b[0]] ?? 0) - a[1] * (WEIGHTS[a[0]] ?? 0)
      )
      .slice(0, n);
  }

  // Called by POST /state/push (browser MediaPipe every 2 s)
  // Accumulates focus_seconds while face is detected and session is active
  pushBrowserFocus(faceDetected, focusScore) {
    const now = Date.now();
    if (this.isActive && this._lastPushMs !== null) {
      const dt = Math.min((now - this._lastPushMs) / 1000, 5); // cap at 5 s gap
      if (faceDetected) this.focusSeconds += dt;
    }
    this._lastPushMs  = now;
    this.faceDetected = Boolean(faceDetected);
    this.focusScore   = Math.max(0, Math.min(100, Number(focusScore)));
  }

  // Called by POST /cv/push (Python bridge every ~1 s)
  updateCV({ ear, mar, head_tilt, nose_ratio, face_detected, blink_rate } = {}) {
    if (ear          !== undefined) this.ear          = Number(ear);
    if (mar          !== undefined) this.mar          = Number(mar);
    if (head_tilt    !== undefined) this.headTilt     = Number(head_tilt);
    if (nose_ratio   !== undefined) this.noseRatio    = Number(nose_ratio);
    if (face_detected !== undefined) this.faceDetected = Boolean(face_detected);
    if (blink_rate   !== undefined) this.blinkRate    = Number(blink_rate);
  }

  snapshot() {
    return {
      session_id:      this.sessionId,
      is_active:       this.isActive,
      focus_score:     Math.round(this.focusScore * 100) / 100,
      counts:          { ...this.counts },
      top_distractors: this.topDistractors(5),
      ear:             this.ear,
      mar:             this.mar,
      head_tilt:       this.headTilt,
      nose_ratio:      this.noseRatio,
      face_detected:   this.faceDetected,
      blink_rate:      this.blinkRate,
      focus_seconds:   Math.round(this.focusSeconds * 10) / 10,
    };
  }
}

export const state = new SessionState();
